### DISCUSSION TITLE
Display Misalignment Issue with Corfu in TTY-Child-Frames

### DISCUSSION THREAD

**Discussion by luluman** [minad/corfu#531] [General]
Thank you for providing this excellent completion package and adding support for tty-child-frames. It's fantastic to have a native popup solution in the terminal, enhancing efficiency and robustness.

I've tested this feature and found a display alignment bug in Corfu. Using the Emacs build from commit [196c761c66](https://git.savannah.gnu.org/cgit/emacs.git/commit/?h=scratch/tty-child-frames&id=196c761c66d47ff6a93d9134ca73d1c08a9125db), the completion popup appears one character to the left of the intended position. I attempted to resolve the issue by setting `(internal-border-width . 0)` but was unable to fix it.

**Steps to Reproduce:**
1. Compile Emacs using commit `196c761c66`.
2. Install the Corfu package and attempt to complete a simple word.

<img width="223" alt="image" src="https://github.com/user-attachments/assets/de48020f-2bd4-40dc-81c3-7e4eff754edb">

**Comment by minad** [comment]
Thank you for testing this with the new tty feature. Obviously this feature is not yet fully supported by Corfu given that we are talking about a separate branch of the Emacs development repository. We will revisit this in the future when child frame tty support has been merged into the Emacs mainline. In the mean time, any help, testing and fixes are appreciated.

**Reply by galeo** [reply]
The tty-child-frames feature has been merged into Emacs' mainline. I did a quick test, and when using tty-child-frames, the alignment issue does occur. This is likely caused by how the frame borders are handled, and it should be relatively easy to fix.

Another consideration is whether corfu should support displaying the frame border in tty-child-frames. By adjusting two parameters in `corfu--frame-parameters`: `(border-width . 1)` and `(undecorated . nil)`, then running the `standard-display-unicode-special-glyphs` command, then corfu will display the frame border drawn using unicode glyphs. This also requires proper handling of the frame border calculations. And more，it may also be worth adjusting corfu's default color faces to better suit tty-child-frames.

**Reply by minad** [reply]
@galeo I am happy to add improvements to Corfu later on, but right now I think it is too early with many changes still happening on Emacs master. Otherwise we would work against each other. As far as I have understood, Gerd is testing his TTY child frames against Corfu and posframe.

**Reply by minad** [reply]
@galeo I am currently improving Corfu TTY support. Do you think using decoration for the Corfu popup is a good idea or does it waste too much screen real estate? I quickly tried it and I am not that convinced. In any case, it should be easy to fix the alignment.

EDIT: The Emacs built-in GUI elements usually don't use a border, e.g., menu-bar-mode or the context menu popup.

**Reply by minad** [reply]
> And more，it may also be worth adjusting corfu's default color faces to better suit tty-child-frames.

Regarding the faces - do you have suggestions? I've tested on TERM=xterm-256color and TERM=xterm-direct (24bit) and there the faces obviously work well. The default Corfu colors are taken from an old version of the Modus themes. I think supporting terminals with fewer colors is difficult.

EDIT: I slightly improved the faces for 8 color terminals in https://github.com/minad/corfu/commit/6d0fe4653987d3d7cfcade2ea72573ab39493ac1.

**Reply by galeo** [reply]
> Do you think using decoration for the Corfu popup is a good idea or does it waste too much screen real estate? I quickly tried it and I am not that convinced.

Decoration doesn’t seem essential for the Corfu popup, and I’m satisfied with how it looks now. 

> I've tested on TERM=xterm-256color and TERM=xterm-direct (24bit) and there the faces obviously work well. The default Corfu colors are taken from an old version of the Modus themes. I think supporting terminals with fewer colors is difficult.

The current color setup is also acceptable for me. Since there’s no border, the corfu-popupinfo and the candidate popup share the same background, and together with the scrollbar and corfu-current background, the overall look can feel a bit cluttered. But it’s still usable, and I guess it just takes some getting used to.

For comparison, here’s a screenshot from Neovim. 

<img width="1114" height="269" alt="neovim-completion-floating-window" src="https://github.com/user-attachments/assets/b366ef24-1595-49cb-8f02-294c4b391ba5" />

Just adding this as a reference. Its completion popup has a border, which makes the visual structure a bit easier to parse. If Corfu could use a border instead of relying solely on background colors to define the boundaries, it might make things visually clearer — even if it costs a bit of space. I experimented with borders before and got a similar result, but I got busy and didn’t follow up on it.

**Reply by minad** [reply]
Okay, I think I would be fine with adding borders if we make it customizable, e.g., `corfu-border-on-tty`. In the screenshot above I would prefer if the main popup and the info popup share a common border, as is the case in Corfu on GUI.

**Reply by galeo** [reply]
A quick test. After adding a border to the popup and removing its background color, here’s the result.

Horizontally stacked:
<img width="1200" height="364" alt="horizontal" src="https://github.com/user-attachments/assets/23aa9369-b00e-480c-bfa5-c1fefca13850" />

Vertically stacked:
<img width="849" height="640" alt="vertical" src="https://github.com/user-attachments/assets/58617434-1bc0-4970-b802-5b27f21d76a2" />

> In the screenshot above I would prefer if the main popup and the info popup share a common border, as is the case in Corfu on GUI.

As shown in the screenshots above, there are gaps between the borders—especially when the two popups are vertically stacked. This has to do with the font used for the border-drawing characters. I don’t know how to make two popups share a common border currently. The popup border characters always include some padding around them, so the border lines don’t join cleanly and there will be visible gaps. I’m not sure whether it’s possible to work around this by manually drawing the borders using characters like ├ ┤ ┬ ┴, but that would get quite complicated.

**Reply by minad** [reply]
> A quick test. After adding a border to the popup and removing its background color, here’s the result.

Happy to take a PR. I hope we don't have to add much more special casing to the already complicated frame position computations?

> As shown in the screenshots above, there are gaps between the borders—especially when the two popups are vertically stacked.

Yes, but theese gaps cannot be avoided right? They are due to the borders drawn in separate lines?

> The popup border characters always include some padding around them, so the border lines don’t join cleanly and there will be visible gaps. I’m not sure whether it’s possible to work around this by manually drawing the borders using characters like ├ ┤ ┬ ┴, but that would get quite complicated.

Right. We could ask Emacs to use joins if child frames overlap, but that's maybe asking too much, and too difficult to implement. Given that Neovim renders the popups also like this, we should be fine with separate frames.

**Reply by galeo** [reply]
> Happy to take a PR. I hope we don't have to add much more special casing to the already complicated frame position computations?

I’ll try to prepare a pull request this weekend.

**Comment by amosbird** [comment]
The completion popup also misbehaves when displayed above the cursor. It leaves visible gap lines between the popup and the cursor. Additionally, when completion is aborted in such cases, the child frame sometimes remains visible, resembling a lingering shadow overlay effect.

emacs: 7ac7ebef6cd *  Update c-ts-mode and c++-ts-mode's verified revision
corfu: bbc205c * Improve docstring

[vid-2024-12-25-120202.webm](https://github.com/user-attachments/assets/61eb5fcb-48a3-4157-bae4-88130da90af9)

**Reply by amosbird** [reply]
It also doesn't work when there are multiple text terminal frames. All frames except the first one that activates the corfu popup will receive this error: 

`Error in post-command-hook (corfu--post-command): (error "Window is dedicated to ‘xxx’")`

The popup's child frame shows garbage content instead.

**Reply by minad** [reply]
I suggest to report also discuss odd behavior upstream as long as it is not clear that Corfu is at fault. The goal is that TTY and GUI child frames behave as consistently as possible.

**Reply by minad** [reply]
@amosbird Could you please provide more reproduction steps? I am currently trying to improve Corfu on TTY. Please try the newest commit from the Corfu repo.

**Reply by amosbird** [reply]
Sure, I’ll try again with the latest commit when I get a moment and see if I can still reproduce it. Thanks!

**Reply by amosbird** [reply]
I’ve tested again with the latest commit, and I can no longer reproduce all the issues. Everything works fine now on my side.

Thank you very much for your hard work on improving Corfu on TTY, really appreciated!
